criu/files-reg.c

```diff
@@ -1726,6 +2044,78 @@ static int rfi_remap(struct reg_file_info *rfi, int *level)
 	return 0;
 }
 
+/*
+ * Compares the file's build-id with the stored value.
+ * Returns 1 if the build-id of the file matches the build-id that was stored
+ * while dumping, -1 if there is a mismatch or 0 if the build-id has not been
+ * stored or could not be obtained.
+ */
+static int validate_with_build_id(const int fd, const struct stat *fd_status,
+					const struct reg_file_info *rfi)
+{
+	unsigned char *build_id;
+	int build_id_size, i;
+
+	if (!rfi->rfe->has_size)
+		return 1;
+
+	if (!rfi->rfe->n_build_id)
+		return 0;
+
+	build_id = NULL;
+	build_id_size = get_build_id(fd, fd_status, &build_id);
+	if (!build_id || build_id_size == -1)
+		return 0;
+
+	if (build_id_size != rfi->rfe->n_build_id) {
+		pr_err("File %s has bad build-ID length %d (expect %d)\n", rfi->path,
+				build_id_size, (int) rfi->rfe->n_build_id);
+		xfree(build_id);
+		return -1;
+	}
+
+	for (i = 0; i < build_id_size; i++)
+		if (build_id[i] != rfi->rfe->build_id[i]) {
+			pr_err("File %s has bad build-ID value %x at index %d (expect %x)\n",
+					rfi->path, build_id[i], i, rfi->rfe->build_id[i]);
+			xfree(build_id);
+			return -1;
+		}
```

Comment:
With the comment above about allocation, this could be:
`if (memcmp(build_id, rfi->rfe->build_id, build_id_size))`